certora/specs/RemoveMarketLiveness.spec

```diff
@@ -0,0 +1,129 @@
+}
+
+// Check that it's possible deallocate expected supply assets, assuming that the IRM doesn't revert and that there is enough liquidity on the market.
+rule canDeallocateExpectedSupplyAssets(env e, bytes data) {
+    Morpho.MarketParams marketParams = Utils.decodeMarketParams(data);
+    Morpho.Id marketId = Utils.id(marketParams);
+    bytes32 id = Utils.unwrapId(marketId);
+    require Morpho.lastUpdate(marketId) == e.block.timestamp, "assume that the IRM doesn't revert";
+
+    uint256 assets = MorphoMarketV1AdapterV2.expectedSupplyAssets(e, marketId);
+
+    require Morpho.totalSupplyAssets(marketId) - assets >= Morpho.totalBorrowAssets(marketId), "assume enough liquidity";
+
+    require MorphoMarketV1AdapterV2.allocation(marketParams) <= max_int256(), "see allocationIsInt256";
+    require MorphoMarketV1AdapterV2.supplyShares(id) <= Morpho.supplyShares(marketId, MorphoMarketV1AdapterV2), "internal accounting of shares is less than actual held shares";
+    require Morpho.supplyShares(marketId, MorphoMarketV1AdapterV2) <= Morpho.totalSupplyShares(marketId), "total supply shares is the sum of the market supply shares";
+    require Morpho.supplyShares(marketId, MorphoMarketV1AdapterV2) < 2^128, "shares fit on 128 bits on Morpho";
+    require assets < 10^32, "safe because market v1 specifies that loan tokens should have less than 1e32 total supply";
+    require Morpho.lastUpdate(marketId) != 0, "assume the market is created";
+    require isAdapter(MorphoMarketV1AdapterV2), "assume the adapter is enabled";
+    require isSentinel(e.msg.sender) || isAllocator(e.msg.sender), "setup the call";
+    require e.msg.value == 0, "setup the call";
+    require marketParams.loanToken == MorphoMarketV1AdapterV2.asset(), "setup the call";
+    require marketParams.irm == MorphoMarketV1AdapterV2.adaptiveCurveIrm(), "setup the call";
+
+    deallocate@withrevert(e, MorphoMarketV1AdapterV2, data, assets);
+
+    assert !lastReverted;
+}
+
+// Check that deallocating expected supply assets puts the expected supply assets to zero.
+rule canPutExpectedSupplyAssetsToZero(env e, bytes data) {
+    Morpho.MarketParams marketParams = Utils.decodeMarketParams(data);
+    Morpho.Id marketId = Utils.id(marketParams);
+    require Morpho.lastUpdate(marketId) == e.block.timestamp, "assume that the IRM doesn't revert";
+
+    uint256 assets = MorphoMarketV1AdapterV2.expectedSupplyAssets(e, marketId);
+
+    deallocate(e, MorphoMarketV1AdapterV2, data, assets);
+
+    assert MorphoMarketV1AdapterV2.expectedSupplyAssets(e, marketId) == 0;
+}
+
+// Check that a deallocation that leaves the expected supply assets to zero removes the market.
+rule deallocatingWithZeroExpectedSupplyAssetsRemovesMarket(env e, bytes data, uint256 assets) {
+    Morpho.MarketParams marketParams = Utils.decodeMarketParams(data);
+    Morpho.Id marketId = Utils.id(marketParams);
+    require Morpho.lastUpdate(marketId) == e.block.timestamp, "assume that the IRM doesn't revert";
```

Comment:
we could have a rule for forceDeallocate that checks that we can always do it

Reply:
Adding this as another rule in the wish list 👍